R/teal_lockfile.R

```diff
@@ -39,50 +39,54 @@ teal_lockfile <- function() {
-    if (inherits(old_plan, "sequential")) {
-      future::plan(future::multisession, workers = 2)
-    }
+  shiny::onStop(function() file.remove(lockfile_path))
 
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  # Both capture.output are not needed if stdout = "|"
+  callr_output <-
+    utils::capture.output( # Needed to suppress: 'Opening fd 1' message
+      utils::capture.output( # Needed to suppress: 'PROCESS 'Rterm', running, pid' output
+        callr::r_bg(
+          func = create_renv_lockfile,
+          args = list(
+            lockfile_path = lockfile_path,
+            opts = options()
+          ),
+          # print results to the console
+          stdout = "",
+          # renv setup is orchestrated by its special S3 objects: renv::settings, renv::config and renv::paths
+          #     `package` parameter include `renv` namespace inside the environment of `func = create_renv_lockfile` fun
+          package = "renv"
+          # default env = NULL # means that callr process will use environmental variables
+          #     from the main session (parent process)
+        ),
+        type = "message"
+      ),
+      type = "output"
+    )
+
+  logger::log_trace("lockfile creation started.")
 }
 
-create_renv_lockfile <- function(close = FALSE, lockfile_path = NULL) {
-  checkmate::assert_flag(close)
+create_renv_lockfile <- function(lockfile_path = NULL, opts) {
   checkmate::assert_string(lockfile_path, na.ok = TRUE)
-  promise <- promises::future_promise({
-    # Below we can not use a file created in tempdir() directory.
-    # If a file is created in tempdir(), it gets deleted on 'then(onFulfilled' part.
-    shiny::onStop(function() file.remove(lockfile_path))
+  checkmate::assert_list(opts)
 
-    renv_logs <- utils::capture.output(
-      renv::snapshot(
-        lockfile = lockfile_path,
-        prompt = FALSE,
-        force = TRUE
-        # type = is taken from renv::settings$snapshot.type()
-      )
-    )
-    if (any(grepl("Lockfile written", renv_logs))) {
-      logger::log_trace("lockfile created successfully.")
-    } else {
-      logger::log_trace("lockfile created with issues.")
-    }
+  options(opts)
 
-    lockfile_path
-  })
-  if (close) {
-    # If the initial backend was only sequential, bring it back.
-    promises::then(promise, onFulfilled = function() {
-      future::plan(future::sequential)
-    })
+  renv_logs <- utils::capture.output(
+    renv::snapshot(
+      lockfile = lockfile_path,
+      prompt = FALSE,
+      force = TRUE
+      # type = is taken from renv::settings$snapshot.type()
+    )
+  )
+  # print() is required for callr::r_bg to track console output
+  if (any(grepl("Lockfile written", renv_logs))) {
+    print("lockfile created successfully.")
+  } else {
+    print("lockfile created with issues.")
```

Comment:
I can't see this message in the console. Does it need `stdout = "|"`?

```r
options(
  teal.log_level = "TRACE",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)
library(scda)
pkgload::load_all("teal")
# pkgload::load_all("teal.slice")

data <- teal_data_module(
  ui = function(id) {
    ns <- NS(id)
    tagList(
      textInput(ns("username"), label = "Username"),
      passwordInput(ns("password"), label = "Password"),
      actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    moduleServer(id, function(input, output, session) {
      eventReactive(input$submit, {
        data <- teal_data() |>
          within(
            {
              logger::log_trace("Loading data")
              ADSL <- scda::synthetic_cdisc_data("latest")$adsl
              ADTTE <- scda::synthetic_cdisc_data("latest")$adtte
              iris <- iris
            },
            username = input$username,
            password = input$password
          )

        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]

        data
      })
    })
  }
)

app <- teal::init(
  data = data,
  modules = modules(
    teal.modules.general::tm_data_table("Data Table"),
    example_module("Example Module", datanames = "ADTTE"),
    module(
      ui = function(id) {
        ns <- NS(id)
        tagList(
          tableOutput(ns("filter_summary"))
        )
      },
      server = function(id, datasets) {
        moduleServer(id, function(input, output, session) {
          output$filter_summary <- renderTable({
            datasets$get_filter_overview(datanames = datasets$datanames())
          })
        })
      }
    )
  ),
  filter = teal_slices(
    teal_slice("ADSL", "SEX"),
    teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
    module_specific = TRUE,
    mapping = list(
      global_filters = "ADSL AGE",
      `Example Module` = "ADSL SEX"
    )
  )
)

shinyApp(app$ui, app$server)

```

Reply:
I see it when I run `app <- teal::init` and I don't run `shinyApp(app$ui, app$server)`.
It seems like once you run `shinyApp(app$ui, app$server)` before lockfile is created, lockfile will not be created.
Once you end app session, you can see `Error: invoke_wrapped: throwing std::runtime_error`
Like `shiny` would block `callr` from running.

Unfortunately I was only testing today till `app <- teal::init` . I have seen it working during shiny runtime yesterday. Will troubleshoot what brought this inconvenience

Reply:
This is the last thing which blocks me from approval. Please investigate if it is possible or remove `print`

Reply:
Yes, will try to figure out why this is not displayed when app is run

Reply:
I just checked that `r_bg` logging is not blocked if shiny app is running. e.g.

```r
library(callr)
rx <- r_bg(function() {Sys.sleep(5); print('Hi!'); renv::snapshot(prompt = FALSE, force = TRUE)}, stdout = "")

app <- init(
  data = teal_data(iris = iris),
  modules = example_module(label = "example teal module")
)

shinyApp(app$ui, app$server)

```

Produces renv.lock file and renv output is pasted in the console. Trying to figure out why it's not happening in teal

Reply:
Hey @gogonzo this is very strange behavior. I think I reached the wall.
I pushed a test version of `create_renv_lockfile_2` to demonstrate one behavior.
When you run `shinyApp()` all `callr` processes keep running, but the ones started inside `teal` get frozen

Try this code on the current branch

```r
library(callr)
rx <- r_bg(function() {lapply(1:15, function(x) {print(paste0('custom: ', x)); Sys.sleep(1)})}, stdout = "")

app <- init(
  data = teal_data(iris = iris),
  modules = example_module(label = "example teal module")
)

shinyApp(app$ui, app$server)
````

Counter for the `r_bg` process started in the main session gets till the end of the counting, but the counter in teal stops. This is just to show that teal-callr process is stopped/killed and the lockfile will never be created when `shinyApp()` is run

<img width="429" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/7fcc2f48-099d-44e7-9dc5-be1fffa9cd01">

Reply:
As you mentioned, I needed to store the process to a variable to prevent it from being killed 027faee48cda8a1c067a0b59d195fb05385f0472